Enable configuration of Puma's thread pool size

mnagy opened the pull request:
Add two configuration variables for determining Puma's thread pool size,
`PUMA_MIN_THREADS` and `PUMA_MAX_THREADS`.

Trello: https://trello.com/c/kvzlecgq/853-5-tune-ruby-based-on-memory-limits-evg

mfojtik commented:
this is missing the tuning magic..

mfojtik commented:
the default values 0, 16 should be based on the # of memory

mnagy commented:
See the trello card for the explanation. Basically, the number of threads
have no impact on memory consumed whatsoever. I have kept Puma's defaults.

On Wed, Mar 2, 2016, 20:27 Michal Fojtik <notifications@github.com> wrote:

> the default values 0, 16 should be based on the # of memory
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/sti-ruby/pull/96#issuecomment-191386310>.
>

bparees commented:
I find that kinda hard to believe.  Were your threads actually doing anything interesting in their request flow?

mfojtik commented:
afaik puma can also pre fork. more memory means more threads ... means more
http throughput.
On Mar 2, 2016 22:46, "Ben Parees" <notifications@github.com> wrote:

> I find that kinda hard to believe. Were your threads actually doing
> anything interesting in their request flow?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/sti-ruby/pull/96#issuecomment-191447266>.
>

mfojtik commented:
@mnagy 

* One worker per core (-w)
* Threads to be determined in connection with RAM availability and application and
* Threads = Connection Pool (more threads, more concurrent connection can be handled)

mnagy commented:
Updated, PTAL.

I have gathered a lot of data and used that to guess the maximum number of workers based on memory limits. I set the number of workers based on number of cores that we have available through cgroups. I parse that using an awk script. The maximum number of workers determined using the memory limit will cap that.

I left the maximum of thread be 16, but I wouldn't mind making that higher. I think the ideal number of threads really depends on the application. But the ideal number of workers should depend only on the number of cores.

mnagy commented:
Btw, I should mention that `export_memory_limit` and `export_number_of_cores` should belong to sti-base. I keep them here in this PR so they can be easily reviewed and tested. They should be reused by other images. Also, I've only updated 2.0 version, I'll copy the changes when this is satisfactorily reviewed and lgtm'd.

bparees commented on 2.0/contrib/etc/puma.cfg:
```diff
@@ -1,4 +1,36 @@
+    cores = ENV.fetch('NUMBER_OF_CORES', '1').to_i
+    per_worker_memory = per_worker_base_memory + per_thread_memory*get_max_threads()
+    max_workers = (get_max_memory() - base_memory) / per_worker_memory
+    [cores, max_workers].min
```
why do we never want more workers than there are cores?  is it not possible for a worker to be blocked on IO, such that other workers could proceed on the same core?

mnagy replied:
I'm following one core one worker recommendation from Puma docs. I'm assuming that threads are dealing with blocking IO.

bparees replied:
ok, can you throw a comment in this logic either linking to the doc or describing the reasoning behind the logic?

mfojtik replied:
it also should be properly documented in official documentation

mnagy replied:
Good point. I'll add a link and update docs.

bparees replied:
still needs doc.

mnagy commented on 2.0/s2i/bin/run:
```diff
@@ -1,5 +1,26 @@
+    return 0
+  fi
+  return 1
+}
```
@mfojtik @bparees do you think these two functions are reasonably good? If yes, then I would like to move these into sti-base and just reuse them here, but I need to at least know this approach in general makes sense to you.

bparees replied:
I think i'd expect them to just return values instead of setting env variables, but otherwise they look ok.

(I guess if they returned values i'm not sure how you'd handle error conditions though, so maybe it's best as is)

mnagy replied:
That's exactly why I chose to use an export and return a 1/0 (and used the "export_" prefix)

mfojtik replied:
@mnagy i am sorry for whoever is going to understand this in the future :-)

mfojtik replied:
@mnagy instead of returning 0/1, just echo the value. The function can be used later as:

```
cores = $(num_of_cores)
[[ "$cores" == "0" ]] && echo "no cores means error"
```

mnagy replied:
@mfojtik yeah, me too :)

@mfojtik I don't want to use special values as errors as it's very messy. I can do with not echoing anything and the user needing to test for zero string. I guess that exporting it explicitly in the script and not inside of the function will make it easier for someone to find where the variable comes from..

mfojtik replied:
@mnagy you have to either check for the return 1 or the `NUMBER_OF_CORES` ;-) also the user of this function will need to know what variable it is exporting, since you hide it in the body (which is not the best api design ;-)

mfojtik replied:
@mnagy after thinking some more about this, I think it would make sense to do the parsing of all limits in python script, something like:

```
export $(python read_limits.py)
```

The `read_limits.py` will return env var key/pairs with all properties... we can include it in sti-base.

mfojtik commented on 2.0/contrib/etc/puma.cfg:
```diff
@@ -1,4 +1,36 @@
+    per_worker_base_memory = 15 * 1024 * 1024
+    per_thread_memory = 103 * 1024 # roughly 0.1 MiB
+
+    cores = ENV.fetch('NUMBER_OF_CORES', '1').to_i
```
i wonder if you can't do the math here (reading the file with ruby and parsing the value... would be much nicer than doing so with awk ;-)

mnagy replied:
I would need to reimplement it in every s2i language, that's why I did it
in awk

On Mon, Mar 14, 2016, 18:46 Michal Fojtik <notifications@github.com> wrote:

> In 2.0/contrib/etc/puma.cfg
> <https://github.com/openshift/sti-ruby/pull/96#discussion_r56043523>:
>
> > +end
> > +
> > +def get_min_threads()
> > +    ENV.fetch('PUMA_MIN_THREADS', '0').to_i
> > +end
> > +
> > +def get_max_threads()
> > +    ENV.fetch('PUMA_MAX_THREADS', '16').to_i
> > +end
> > +
> > +def get_workers()
> > +    base_memory = 50 * 1024 * 1024
> > +    per_worker_base_memory = 15 * 1024 * 1024
> > +    per_thread_memory = 103 * 1024 # roughly 0.1 MiB
> > +
> > +    cores = ENV.fetch('NUMBER_OF_CORES', '1').to_i
>
> i wonder if you can't do the math here (reading the file with ruby and
> parsing the value... would be much nicer than doing so with awk ;-)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/sti-ruby/pull/96/files#r56043523>.
>

mnagy commented:
Rewritten to use sti-base's python utility, added docs to the README.md. Still only for one version, will copy it to the other one after review.
PTAL

bparees commented on 2.0/contrib/etc/puma.cfg:
```diff
@@ -1,4 +1,36 @@
+def get_max_memory()
+    if ENV.has_key?('MEMORY_LIMIT_IN_BYTES')
+        return ENV['MEMORY_LIMIT_IN_BYTES'].to_i
+    end
```
should you be checking for the new no limit value now?

mnagy replied:
no, there's no new no limit value, we just added maximum. We don't mind very high values here, because we're maxing out on the number of cores anyway

bparees commented on 2.0/s2i/bin/run:
```diff
@@ -1,5 +1,26 @@
 #!/bin/bash
 
+function export_memory_limit() {
```
this can go away now right?

bparees commented on 2.0/s2i/bin/run:
```diff
@@ -1,5 +1,26 @@
+  return 1
+}
+
+function export_number_of_cores() {
```
should go away?

mnagy commented:
@bparees I accidentally pushed to a wrong branch, sorry :( PTAL

mfojtik commented on 2.0/s2i/bin/run:
```diff
@@ -10,6 +10,16 @@ set -e
 export RACK_ENV=${RACK_ENV:-"production"}
 
 if is_puma_installed; then
+  export_str=$(cgroup-limits.py)
```
i think it is safe to do just `export $(cgroup-limits.py)` now because it will never be empty (it will print the maximum memory ;-)

mnagy replied:
no, it is not, if cgroup-limits.py fails (and throws a traceback) the export command would still pass and run with the garbage output. Tested.

mfojtik replied:
how the check below help with testing that situation :-)

mnagy replied:
The check below is for something else (will not use it). But the fact that it's on a separate line is the difference. `export_str=$(cgroup-limits.py) ; export $export_str` will fail and the script will exit if cgroup-limits.py fails (because of `set -e`). If you try `export $(cgroup-limits.py)` it will ignore cgroup-limits.py failure and will try to work with it's output anyway.

mfojtik commented on 2.0/s2i/bin/run:
```diff
@@ -10,6 +10,16 @@ set -e
+  export_str=$(cgroup-limits.py)
+  [[ -n "$export_str" ]] && export $export_str
+
+  if [[ -z "${MEMORY_LIMIT_IN_BYTES:-}" ]]; then
```
this will never be empty, right?

mnagy replied:
if it is empty it means cgroup-limits.py has failed, better warn the user.

mfojtik commented on 2.0/s2i/bin/run:
```diff
@@ -10,6 +10,16 @@ set -e
+    echo "Warning: Can't detect memory limit"
+  fi
+  if [[ -z "${NUMBER_OF_CORES:-}" ]]; then
+    echo "Warning: Can't detect number of cores"
```
I would prefer to not display these warnings... users might be concerned but there is nothing to be concerned about...
If this var is empty, we just default to some default value.

mnagy replied:
we do but it might mean that something went wrong in the script. Unless there's a case where cgroups are not used. I don't think the warning would be harmful in that case, I assume it'll be a minority issue?

mfojtik replied:
@mnagy if the script error out, we should display the traceback and exit. if the script return nothing (that will never happen, because we always print the max memory var), we just export it. if the vars are empty (never happen because the files are always there with some default value) it means the script failed and we will never get here :-)

mnagy replied:
If the script errors out, it will exit. But yeah, now that we print the maximum value we can assume it'll never be empty. The script will ignore missing cgroup files, that's why the warning is there. If we want to assume they will always be there, then we can remove the warnings and make the cgroup script not ignore them. Do we want to do that, though?

bparees replied:
Maybe just update the warning with a little more info, like "Can't detect number of cores, defaulting setting FOO to BAR"

that way we still get the warning (for when we are debugging user environments) but users who see it will have a warm fuzzy about the implication of the warning.

You can even print "Warning: can't detect number of cores.  Setting FOO to BAR, set env variable XXX to override"

(and we should only print the warning if they haven't already explicitly set the values that would be set based on the cgroup value.  That is, if we weren't going to do anything with the cgroup value anyway, then it's not worth warning them about it)

mnagy commented:
PTAL. Warnings are now handled in the cgroup script, so this makes this PR a bit simpler. I have updated 2.2 as well, this should now be ready for merge, if it gets approved.

mfojtik commented on 2.0/contrib/etc/puma.cfg:
```diff
@@ -1,4 +1,40 @@
+def get_max_memory()
+    if ENV.has_key?('MEMORY_LIMIT_IN_BYTES')
+        return ENV['MEMORY_LIMIT_IN_BYTES'].to_i
```
cool kinds write this as:

```
return ENV['MEMORY_LIMIT_IN_BYTES'].to_i if ENV.has_key? "MEMORY_LIMIT_IN_BYTES"
```

mfojtik commented on 2.0/s2i/bin/run:
```diff
@@ -10,6 +10,9 @@ set -e
 
 if is_puma_installed; then
+  export_vars=$(cgroup-limits.py)
+  export $export_vars
```
how about:

```
limit_vars=$(cgroup-limits.py) && export $limit_vars
```

it should work:

```
Σ~/.../openshift/origin ∎ t=$(false) && echo 1
Σ~/.../openshift/origin ∎ t=$(true) && echo 1
1
```

mnagy replied:
That won't exit if cgroup-limits.py fails (set -e) which is what I want

mnagy replied:
If you want it to be on one line, I can do `export_vars=$(cgroup-limits.py) ; export $export_vars`

mnagy commented:
Addressed comments, PTAL

bparees commented on 2.0/contrib/etc/puma.cfg:
```diff
@@ -1,4 +1,38 @@
-threads     0, 16
-workers     0
+threads     get_min_threads(), get_max_threads()
+workers     get_workers()
```
hm.  shouldn't i be able to override and explicitly set the number of workers?

mnagy replied:
We can add that, sure.

mnagy commented:
PTAL

bparees commented on 2.0/README.md:
````diff
@@ -135,3 +143,33 @@ docker exec -it <CONTAINER_ID> /bin/bash
+$ docker run -e PUMA_MAX_THREADS=32 --memory=300m -p 8080:8080 sinatra-app
+```
+If memory is more limiting then the number of available cores, the number of
+workers is scaled down accordingly to fit the above formula.
````
"The number of workers can also be set explicilty by setting PUMA_WORKERS."

mnagy replied:
done

bparees commented:
lgtm, will wait for tests to pass.

rhcarvalho commented on 2.0/README.md:
````diff
@@ -135,3 +143,34 @@ docker exec -it <CONTAINER_ID> /bin/bash
+```
+The number of workers is also limited by the memory limit that is enforced using
+cgroups. The builder image assumes that you will need 50 MiB as a base and
+another 15 MiB for every worker thread plus 0.1 MiB for each thread. Note that
````
> every worker thread

The worker is a process, isn't it?

I'd expect this to be "x MiB as a base, another y MiB for every worker process, plus z MiB for each thread."

mnagy replied:
Correct, good catch.

rhcarvalho commented on 2.2/README.md:
````diff
@@ -135,3 +143,34 @@ docker exec -it <CONTAINER_ID> /bin/bash
+50 + 15 * WORKERS + 0.1 * WORKERS * PUMA_MAX_THREADS
+```
+You can specify a memory limit using the
+`--memory` flag:
````
this probably don't need to be in multiple lines :dancer:

rhcarvalho commented on 2.0/contrib/etc/puma.cfg:
```diff
@@ -1,4 +1,40 @@
+def get_max_memory()
+    return ENV['MEMORY_LIMIT_IN_BYTES'].to_i if ENV.has_key? 'MEMORY_LIMIT_IN_BYTES'
+
+    # Assume unlimited memory. 0.size is the number of bytes a ruby
```
"Ruby", to be consistent with the next occurrence below? ;)

rhcarvalho commented on 2.0/contrib/etc/puma.cfg:
```diff
@@ -1,4 +1,40 @@
+
+    base_memory = 50 * 1024 * 1024
+    per_worker_base_memory = 15 * 1024 * 1024
+    per_thread_memory = 103 * 1024 # roughly 0.1 MiB
```
What?! 103 looks really weird, just like these dancers :dancers:

mnagy replied:
Very well. 128 it is. (I took a quick glance at the data and 128 should be all right)

rhcarvalho commented on 2.0/s2i/bin/run:
```diff
@@ -10,6 +10,8 @@ set -e
 export RACK_ENV=${RACK_ENV:-"production"}
 
 if is_puma_installed; then
+  export_vars=$(cgroup-limits.py) ; export $export_vars
```
We should probably not have the ".py" suffix in the file name, and simple run it as an executable in the PATH...
There are other utils coming from sti-base that have no extension in the file name.

bparees replied:
updating sti-base is going to be pretty painful and delay this.  do you really think this is critical?

rhcarvalho replied:
Not critical, but was something I wanted to have mentioned in the original PR to sti-base but had no chance to review before it merged ;)

If we'll ever fix it, better do it sooner than later, before this name will be all around every sti-* image.

mnagy replied:
Let's just do it now, will file a PR.

rhcarvalho commented:
@mnagy in terms of error handling, what happens if PUMA_MIN_THREADS (and others) is not a number?

Will it stop the build? Show a stack trace? Both? Something else?

mnagy commented:
@rhcarvalho I actually had a check for that, but lost it. Found it in an older branch :) Thanks for catching that.

mnagy commented:
Addressed @rhcarvalho's comments.

bparees commented:
lgtm.  test will presumably fail until new sti-base image is pushed.

mfojtik commented on 2.0/s2i/bin/run:
```diff
@@ -7,9 +7,21 @@ function is_puma_installed() {
 
 set -e
 
+function check_number() {
```
`check_positive_number` ;-)

mnagy replied:
It's checking for a non-negative number, not a positive number.

mfojtik commented:
small renaming nit, but lgtm :-)

mfojtik commented on 2.0/s2i/bin/run:
```diff
@@ -7,9 +7,21 @@ function is_puma_installed() {
 export RACK_ENV=${RACK_ENV:-"production"}
 
 if is_puma_installed; then
+  export_vars=$(cgroup-limits) ; export $export_vars
```
@mnagy can you check what happen when you set all the vars above to 0? will puma start?

mnagy replied:
Yes, it will start. If you set max threads to 0, it won't respond to any requests, however.

rhcarvalho commented on 2.0/s2i/bin/run:
```diff
@@ -7,9 +7,21 @@ function is_puma_installed() {
+}
+check_number PUMA_WORKERS     "${PUMA_WORKERS:-0}"
+check_number PUMA_MIN_THREADS "${PUMA_MIN_THREADS:-0}"
+check_number PUMA_MAX_THREADS "${PUMA_MAX_THREADS:-0}"
```
Wouldn't it be better/easier to do this in Ruby?

I wouldn't expect Bash to have anything to do with this vars...

We simply launch Puma, it runs its puma.cfg script, and if the script fails the container/pod fails, etc...
If the env vars are present and valid, use them, if invalid, error, if not present, use defaults.

mnagy replied:
We discussed this off-line. My main reason for leaving it here is that I think of the env vars as an interface to our image and we do validate vars in bash for example in the MySQL image (and I assume other as well).

bparees commented:
[test]

mnagy commented:
Small update. I've updated the documentation on the formula. Changed 0.1 MiB to 128 KiB and 0.125 MiB where appropriate.

mfojtik commented:
LGTM
